src/ZkFold/Base/Protocol/Plonkup/PlonkConstraint.hs

```diff
@@ -34,41 +33,39 @@ data PlonkConstraint i a = PlonkConstraint
     , qr :: a
     , qo :: a
     , qc :: a
-    , x1 :: Maybe (Var (Vector i))
-    , x2 :: Maybe (Var (Vector i))
-    , x3 :: Maybe (Var (Vector i))
+    , x1 :: Var a (Vector i)
+    , x2 :: Var a (Vector i)
+    , x3 :: Var a (Vector i)
     }
     deriving (Show, Eq)
 
-instance (Arbitrary a, Binary a, KnownNat i) => Arbitrary (PlonkConstraint i a) where
+instance (Ord a, Arbitrary a, Binary a, KnownNat i) => Arbitrary (PlonkConstraint i a) where
     arbitrary = do
         qm <- arbitrary
         ql <- arbitrary
         qr <- arbitrary
         qo <- arbitrary
         qc <- arbitrary
-        k <- elements [1, 2, 3]
-        xs0 <- sort <$> replicateA k (Just . NewVar . toByteString @a <$> arbitrary)
-        let (x, y, z) = case replicate (3 -! k) Nothing ++ xs0 of
-              [x', y', z'] -> (x', y', z')
-              _            -> error "impossible"
-        return $ PlonkConstraint qm ql qr qo qc x y z
+        let arbitraryNewVar = SysVar . NewVar . toByteString @a <$> arbitrary
```

Comment:
Is it okay if arbitrary x1,x2,x3 are only new assigned variables, not input or constant variables?

Reply:
In particular no `ConstVar one`s are generated...

Reply:
@vlasin

Reply:
I think it is fine. In general, I think we should get rid of this instance and just use Arbitrary for ArithmeticCircuit.